### PR TITLE
Warm_Up -Week3

### PR DESCRIPTION
**Pull request by bhumikama**
[Warmup requirements](https://github.com/HackYourFuture-CPH/node.js/tree/main/week3/homework#warmup)

Demonstration of  SQL injection
Fix SQL injection vulnerability
[Video demonstrating SQL injection vulnerability](https://www.loom.com/share/27746b9288e3424eb40b95ef358fb4c5)

### CODE REVIEW

File: nodejs/week3/app.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+//     console.error(e);
+//     res.status(500).json({ error: "Internal server error" });
+//   }
+// });
+
+//http://localhost:3000/api/contacts?sort=first_name%20DESC
+//http://localhost:3000/api/contacts?sort=first_name%20DESC; DROP TABLE contacts;
+
+contactsAPIRouter.get("/", async (req, res) => {
+  let query = knexInstance.select("*").from("contacts");
```
Comment (urbanogilson) [inline nodejs/week3/app.js]:
```suggestion
  let query = knexInstance("contacts");
```

File: nodejs/week3/app.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  let validSortFields = ["first_name", "last_name", "email", "phone"];
+  let validSortOrder = ["ASC", "DESC"];
+
+  if ("sort" in req.query) {
+    const sortParameter = req.query.sort.toString();
+    let parts = sortParameter.split(" ");
+    let sortField = parts[0];
+    let sortOrder = parts[1] ? parts[1].toUpperCase() : "ASC"; // adding a default value if not provided
```
Comment (urbanogilson) [inline nodejs/week3/app.js]:
```suggestion
    let sortOrder = parts[1]?.toUpperCase() || "ASC"; // adding a default value if not provided
```

File: nodejs/week3/app.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+contactsAPIRouter.get("/", async (req, res) => {
+  let query = knexInstance.select("*").from("contacts");
+  let validSortFields = ["first_name", "last_name", "email", "phone"];
+  let validSortOrder = ["ASC", "DESC"];
+
+  if ("sort" in req.query) {
+    const sortParameter = req.query.sort.toString();
+    let parts = sortParameter.split(" ");
+    let sortField = parts[0];
```
Comment (urbanogilson) [inline nodejs/week3/app.js]: Make it case-insensitive so you can accept filters like `First_Name`

```suggestion
    let sortField = parts[0].toLowerCase();
```

File: nodejs/week3/app.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+app.use(express.json());
+
+const apiRouter = express.Router();
+app.use("/api", apiRouter);
+
+const contactsAPIRouter = express.Router();
+apiRouter.use("/contacts", contactsAPIRouter);
+
+// contactsAPIRouter.get("/", async (req, res) => {
```
Comment (urbanogilson) [inline nodejs/week3/app.js]: You can remove the commented code

**Review by urbanogilson — COMMENTED** [review]
Good job!

Very good video, clear and well explained. Congratulations!

**Comment by bhumikama** [comment]
Thanks @urbanogilson for your inputs and review.
I have included the suggested changes . Please have another look.

File: nodejs/week3/app.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    if (
+      !validSortFields.includes(sortField) ||
+      !validSortOrder.includes(sortOrder)
+    ) {
+      return res.status(404).json({ message: "invalid sort parameter" });
+    }
+    query = query.orderBy(sortField, sortOrder); // after validating apply sql query
+  }
+  try {
```
Comment (urbanogilson) [inline nodejs/week3/app.js:45]: I would move this line to line 27 to make all the code inside the try catch

**Review by urbanogilson — APPROVED** [review]

**Pull request merged**